core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/ProcessUtil.java

```diff
@@ -126,6 +130,83 @@ public static Process executeAndWait(
     return process;
   }
 
+  /**
+   * Start the process, writing the stdout of process to {@code stdout}, stderr of process to {@code
+   * stderr}, and terminate process by {@code timeout}.<br>
+   *
+   * @param commandLine arguments of process command
+   * @param timeout timeout for process. If process duration > {@code timeout} than kill process and
+   *     throw {@link TimeoutException}.
+   * @param timeUnit timeUnit of the {@code timeout}.
+   * @param stdout a consumer where stdout will be redirected
+   * @param stderr a consumer where stderr will be redirected
+   * @return the started process
+   * @throws InterruptedException in case terminate process
+   * @throws IOException in case I/O error, or command execution error, or stderr is not empty after
+   *     the command execution
+   * @throws TimeoutException if process gets more time then defined by {@code timeout}
+   */
+  public static Process executeAndWait(
+      String[] commandLine,
+      int timeout,
+      TimeUnit timeUnit,
+      ListLineConsumer stdout,
+      ListLineConsumer stderr)
+      throws TimeoutException, IOException, InterruptedException {
+    ProcessBuilder pb = new ProcessBuilder(commandLine);
+
+    Process process = pb.start();
+
+    CompletableFuture<Void> future =
+        CompletableFuture.runAsync(
+            () -> {
+              try {
+                // consume logs until process ends
+                process(process, stdout, stderr);
+              } catch (IOException e) {
+                LOG.error(
+                    format(
+                        "Failed to complete reading of the process '%s' output due to occurred error",
+                        Joiner.on(" ").join(commandLine)),
+                    e);
+              }
+            });
+
+    if (!process.waitFor(timeout, timeUnit)) {
+      try {
+        ProcessUtil.kill(process);
+      } catch (RuntimeException x) {
+        LOG.error("An error occurred while killing process '{}'", Joiner.on(" ").join(commandLine));
+      }
+      throw new TimeoutException(
+          format(
+              "Process '%s' was terminated by timeout %s %s.",
+              Joiner.on(" ").join(commandLine), timeout, timeUnit.name().toLowerCase()));
+    }
+
+    future.join(); // wait until process output is read
+
+    if (process.exitValue() == 0) {
+      return process;
+    }
+
+    String errorMessage = stderr.getText();
```

Comment:
So what? How are you going to handle 10GB of error text here?

Reply:
I propose to cut error text which we is setting as error message of exception, to 1000 symbols.

Reply:
As far as I can see code that you proposed is not enough. Because OOM will happen earlier. Do you see that?

Reply:
It could happen when we are writing into the ListLineConsumer's **stdout** and **stderr** here:
- https://github.com/eclipse/che/blob/61c6247e26d6ee38cc8f029a2eb41135eeb4b216/core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/ProcessUtil.java#L56
- https://github.com/eclipse/che/blob/61c6247e26d6ee38cc8f029a2eb41135eeb4b216/core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/ProcessUtil.java#L59

Reply:
So this method designed for commands that for sure are not producing some large output?
What if you change ListLineConsumer's to LineConsumer and remove part of the stderr from output.

Reply:
It's possible with some remarks https://github.com/eclipse/che/pull/8652#discussion_r167246019

Reply:
Not clear enough.  Is this method designed to handle large output for stderror or stdout?

Reply:
ListLineConsumer is not designed to avoid OOM when something write 10 GB by using method writeLine().